File: content/manual/v8.2/source/includes/aggregation/aggregation-examples/unpack-arrays/full-files/UnpackArrays.kt
package org.example

import com.mongodb.client.model.Accumulators
import com.mongodb.client.model.Aggregates
import com.mongodb.client.model.Field
import com.mongodb.client.model.Filters
import com.mongodb.kotlin.client.coroutine.MongoClient
import kotlinx.serialization.Serializable
import org.bson.Document
import org.bson.conversions.Bson

// start-data-classes
@Serializable
data class Order(
    val orderID: Float,
    val products: List<Product>
)

@Serializable
data class Product(
    val prodID: String,
    val name: String,
    val price: Int
)
// end-data-classes

suspend fun main() {
    val uri = "<connection string>"

    MongoClient.create(uri).use { mongoClient ->
        val aggDB = mongoClient.getDatabase("agg_tutorials_db")
        // start-insert-orders
        val orders = aggDB.getCollection<Order>("orders")
        orders.deleteMany(Filters.empty())

        orders.insertMany(
            listOf(
                Order(
                    6363763262239f, listOf(
                        Product("abc12345", "Asus Laptop", 431),
                        Product("def45678", "Karcher Hose Set", 22)
                    )
                ),
                Order(
                    1197372932325f, listOf(
                        Product("abc12345", "Asus Laptop", 429)
                    )
                ),
                Order(
                    9812343774839f, listOf(
                        Product("pqr88223", "Morphy Richards Food Mixer", 431),
                        Product("def45678", "Karcher Hose Set", 21)
                    )
                ),
                Order(
                    4433997244387f, listOf(
                        Product("def45678", "Karcher Hose Set", 23),
                        Product("jkl77336", "Picky Pencil Sharpener", 1),
                        Product("xyz11228", "Russell Hobbs Chrome Kettle", 16)
                    )
                )
            )
        )
        // end-insert-orders

        val pipeline = mutableListOf<Bson>()

        // start-unwind
        pipeline.add(
            Aggregates.unwind("\$${Order::products.name}")
        )
        // end-unwind

        // start-match
        pipeline.add(
            Aggregates.match(
                Filters.gt("${Order::products.name}.${Product::price.name}", 15)
            )
        )
        // end-match

        // start-group
        pipeline.add(
            Aggregates.group(
                "\$${Order::products.name}.${Product::prodID.name}",
                Accumulators.first("product", "\$${Order::products.name}.${Product::name.name}"),
                Accumulators.sum("total_value", "\$${Order::products.name}.${Product::price.name}"),
                Accumulators.sum("quantity", 1)
            )
        )
        // end-group

        // start-set
        pipeline.add(Aggregates.set(Field("product_id", "\$_id")))
        // end-set

        // start-unset
        pipeline.add(Aggregates.unset("_id"))
        // end-unset

        // start-run-agg
        val aggregationResult = orders.aggregate<Document>(pipeline)
        // end-run-agg

        aggregationResult.collect { println(it) }
    }
}

File: content/manual/v8.2/source/includes/aggregation/aggregation-examples/unpack-arrays/full-files/UnpackArrays.scala
package org.example

import org.mongodb.scala.MongoClient
import org.mongodb.scala.bson.Document
import org.mongodb.scala.model.{Accumulators, Aggregates, Field, Filters}

object UnpackArrays {
  def main(args: Array[String]): Unit = {

    val uri = "<connection string>"
    val mongoClient = MongoClient(uri)
    Thread.sleep(1000)

    val aggDB = mongoClient.getDatabase("agg_tutorials_db")

    // start-insert-orders
    val orders = aggDB.getCollection("orders")

    orders.deleteMany(Filters.empty()).subscribe(
      _ => {},
      e => println("Error: " + e.getMessage),
    )

    orders.insertMany(Seq(
      Document(
        "order_id" -> 6363763262239L,
        "products" -> Seq(
          Document(
            "prod_id" -> "abc12345",
            "name" -> "Asus Laptop",
            "price" -> 431
          ),
          Document(
            "prod_id" -> "def45678",
            "name" -> "Karcher Hose Set",
            "price" -> 22
          )
        )
      ),
      Document(
        "order_id" -> 1197372932325L,
        "products" -> Seq(
          Document(
            "prod_id" -> "abc12345",
            "name" -> "Asus Laptop",
            "price" -> 429
          )
        )
      ),
      Document(
        "order_id" -> 9812343774839L,
        "products" -> Seq(
          Document(
            "prod_id" -> "pqr88223",
            "name" -> "Morphy Richards Food Mixer",
            "price" -> 431
          ),
          Document(
            "prod_id" -> "def45678",
            "name" -> "Karcher Hose Set",
            "price" -> 21
          )
        )
      ),
      Document(
        "order_id" -> 4433997244387L,
        "products" -> Seq(
          Document(
            "prod_id" -> "def45678",
            "name" -> "Karcher Hose Set",
            "price" -> 23
          ),
          Document(
            "prod_id" -> "jkl77336",
            "name" -> "Picky Pencil Sharpener",
            "price" -> 1
          ),
          Document(
            "prod_id" -> "xyz11228",
            "name" -> "Russell Hobbs Chrome Kettle",
            "price" -> 16
          )
        )
      )
    )).subscribe(
      _ => {},
      e => println("Error: " + e.getMessage),
    )
    // end-insert-orders

    Thread.sleep(1000)

    val pipeline = Seq(
      // start-unwind
      Aggregates.unwind("$products"),
      // end-unwind

      // start-match
      Aggregates.filter(Filters.gt("products.price", 15)),
      // end-match

      // start-group
      Aggregates.group(
        "$products.prod_id",
        Accumulators.first("product", "$products.name"),
        Accumulators.sum("total_value", "$products.price"),
        Accumulators.sum("quantity", 1)
      ),
      // end-group

      // start-set
      Aggregates.set(Field("product_id", "$_id")),
      // end-set

      // start-unset
      Aggregates.unset("_id")
      // end-unset
    )

    // start-run-agg
    orders.aggregate(pipeline)
      .subscribe((doc: Document) => println(doc.toJson()),
        (e: Throwable) => println(s"Error: $e"))
    // end-run-agg

    Thread.sleep(1000)
    mongoClient.close()
  }
}
